### DISCUSSION TITLE
Does Gickup push to destinations with force? Is it configurable?

### DISCUSSION THREAD

**Discussion by berezovskyi** [cooperspencer/gickup#144] [Q&A]
Hello, I got first interested in backing up Git repos after Neo4j removed a bunch of code from their repos by force-pushing a nearly-clean branch (though, it seems it was undone at some point as I can see the old code in the 3.x branches now).

Anyway, I set up Gogs and later Gitea to mirror the repos and at some point a "full mirror" option was introduced, which is good if you wish to mirror a codebase that does rebasing in mirrored branches but less than optimal for durable backups. See [this issue](https://github.com/go-gitea/gitea/issues/14076) for more discussion.

I looked [through](https://cooperspencer.github.io/gickup-documentation/docs/configuration/destination/) the [docs](https://cooperspencer.github.io/gickup-documentation/docs/configuration/miscellaneous/) but didn't find any mention of the Gickup behaviour if non-fast-forward changes are pushed to upstream.

Thanks in advance!

**Comment by cooperspencer** [comment]
Hi,
normally gickup will try to pull for a few times, 5 times to be exact, and if it still can't pull, it will delete the repository and clone it again.
Should something like the Neo4j issue have occurred, it would have pulled the forced changes. There is an option to clone it as a backup, which you can store wherever you like and restore if necessary.

Implementing an option to disable the default behavior of gickup sounds interesting and I will look into that.

**Reply by berezovskyi** [reply]
Thanks! I figured out to make it work somewhat according to my needs: gickup to a local dir w/o zip and then apply a backup solution like borg/restic/etc. I think the timestamp in the path could be a problem though.

Regarding a non-forced push, I think rebasing is very common on feature branches. To avoid too many false positives,that feature would also need a support for branch filtering, e.g. `(master|main|devel|maint-.+)` (or an exclusion rule like `(b|pr)?\d+-.+`). I am new to Go but could give a try with a PR for one of the small issues first, with  your guidance.

**Reply by cooperspencer** [reply]
I created a repository and pushed changes to it, then I initialized a second one and force-pushed the new repo over the old one. That's basically what happened with Neo4j?
Then I tried the force parameter in the `go-git` module but I get the same error with or without that parameter, `non-fast-forward update`.

Could be a bug in the module

**Reply by cooperspencer** [reply]
I just came up with an idea. From all the research I did so far, one solution was to rename the branch, fetch all changes and reset the repo.
```bash
git checkout master
git branch new-branch-to-save-current-commits
git fetch --all
git reset --hard origin/master
```

What if I just rename the existing branches with `timestamp-branchname` to keep all changes?

**Reply by cooperspencer** [reply]
https://github.com/cooperspencer/gickup/pull/145/files#diff-28b5602b66ba5bfacacd48c30423065bf23d616dc8a61b3256be1b120f3fc166
these are my thoughts so far.
